### PR TITLE
Import local node

### PR DESCRIPTION
**Pull request by michallowasrzechonek-silvair**

### CODE REVIEW

File: mesh/node.c [hunk 1]
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
...
+	return true;
+}
+
+static bool parse_imported_iv_index(json_object *jobj, uint32_t *idx,
+								bool *update)
+{
+	int tmp;
+	json_object *jvalue;
+
+	if (!json_object_object_get_ex(jobj, "IVindex", &jvalue))
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: see `mesh-db.c` function `get_int` - it might be a good idea to extract this for reuse

File: mesh/node.c [hunk 2]
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
...
+
+	str = (char *)json_object_get_string(jvalue);
+
+	if (sscanf(str, "%04hx", unicast) != 1)
+		return false;
+
+	return true;
+}
+
+static bool parse_imported_device_key(json_object *jobj, uint8_t key_buf[16])
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: same, `mesh_db_read_device_key`/`mesh_db_read_token` - all of these functions read a hexstring from json

File: mesh/node.c [hunk 3]
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
...
+			return false;
+
+		jidx = json_object_iter_peek_name(&iter);
+
+		if (!*jidx)
+			return false;
+
+		switch (strlen(jidx)) {
+
+		case 4: {
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: I don't think you need braces here

File: mesh/node.c [hunk 4]
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
...
+
+	if (!IS_UNASSIGNED(db_node->unicast) &&
+			!mesh_net_register_unicast(net, db_node->unicast,
+			node_get_num_elements(node))) {
+
+		node_remove(node);
+		return false;
+	}
+
+	if (node_uuid_get(node))
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: I don't think this `if` is needed, node objects always have uuids set.

File: mesh/node.c [hunk 5]
```diff
@@ -1672,13 +2040,108 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
 					req, l_free);
 }
 
+void node_import(const char *sender, void *json_data, const uint8_t *uuid,
+			node_ready_func_t cb, void *user_data)
+{
+	struct mesh_node *node;
+	struct mesh_net *net;
+	struct mesh_db_node db_node;
+	struct keyring_net_key *net_keyring;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Unused variable?

File: mesh/node.c [hunk 6]
```diff
@@ -1561,29 +1922,36 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
...
-		l_getrandom(dev_key, sizeof(dev_key));
-		l_getrandom(net_key.old_key, sizeof(net_key.old_key));
-		net_key.net_idx = PRIMARY_NET_IDX;
-		net_key.phase = KEY_REFRESH_PHASE_NONE;
-
-		if (!add_local_node(node, DEFAULT_NEW_UNICAST, false, false,
-						DEFAULT_IV_INDEX, dev_key,
-						PRIMARY_NET_IDX,
-						net_key.old_key))
+		l_getrandom(dev_key, 16);
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: I would leave the `sizeof` and use that in other places too, instead of hardcoding `16`

File: mesh/node.c [hunk 7]
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
...
+	if (json_object_get_type(jobject) != json_type_object)
+		return NULL;
+
+	keys = l_queue_new();
+
+	iter = json_object_iter_begin(jobject);
+	iend = json_object_iter_end(jobject);
+
+	while (!json_object_iter_equal(&iter, &iend)) {
+		net_keyring = l_new(struct keyring_net_key, 1);
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `mesh/node.c:1483:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

File: mesh/node.c [hunk 8]
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
...
+
+	if (node_uuid_get(node))
+		mesh_net_id_uuid_set(net, node_uuid_get(node));
+
+	return true;
+}
+
+static void add_keys(void *net_keys, void *_node)
+{
+	l_info("ADD KEYS CB");
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `mesh/node.c:1725:2: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

File: mesh/node.c [hunk 9]
```diff
@@ -1550,7 +1911,7 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 	} else {
 		/* Callback for create node request */
 		node_ready_func_t cb = req->cb;
-		struct keyring_net_key net_key;
+		net_key = l_new(struct keyring_net_key, 1);
 		uint8_t dev_key[16];
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `mesh/node.c:1915:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

File: mesh/node.c [hunk 10]
```diff
@@ -1672,13 +2040,108 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
...
+
+	if (node)
+		node_remove(node);
+}
+
 void node_create(const char *app_path, const char *sender, const uint8_t *uuid,
 					node_ready_func_t cb, void *user_data)
 {
 	struct managed_obj_request *req;
+	struct mesh_node *node;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `mesh/node.c:2140:20: error: variable ‘node’ set but not used [-Werror=unused-but-set-variable]`

File: mesh/node.c [hunk 11]
```diff
@@ -1426,6 +1785,8 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 	struct managed_obj_request *req = user_data;
 	const char *path;
 	struct mesh_node *node = NULL;
+	struct l_queue *net_keys;
+	struct keyring_net_key *net_key;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `mesh/node.c:1953:3: error: ‘net_key’ may be used uninitialized in this function [-Werror=maybe-uninitialized]`, initialize to `NULL`

File: mesh/node.c [hunk 12]
```diff
@@ -1426,6 +1785,8 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 	struct managed_obj_request *req = user_data;
 	const char *path;
 	struct mesh_node *node = NULL;
+	struct l_queue *net_keys;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `mesh/node.c:1949:5: error: ‘net_keys’ may be used uninitialized in this function [-Werror=maybe-uninitialized]`

File: mesh/node.c (hunk 3 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: Removed

File: mesh/node.c (hunk 4 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: Removed

File: mesh/node.c (hunk 5 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: Correct

File: mesh/node.c (hunk 8 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: log removed

File: mesh/node.c (hunk 1 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: Reading json will be simplified in the near future

File: mesh/node.c (hunk 2 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: Reading json will be simplified in the near future

File: doc/mesh-api.txt [hunk 13]
```diff
@@ -151,7 +151,7 @@ Methods:
 			org.bluez.mesh.Error.InvalidArguments
 			org.bluez.mesh.Error.AlreadyExists,
 
-	 uint64 token ImportLocalNode(string json_data)
+	 uint64 token ImportLocalNode(string json_data, array{byte}[16] uuid)
```
Comment (michallowasrzechonek-silvair) [inline doc/mesh-api.txt]: Update API documentation

File: mesh/mesh.c [hunk 14]
```diff
@@ -758,12 +759,56 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
 
 	l_queue_push_tail(pending_queue, pending_msg);
 
-	node_create(app_path, sender, uuid, create_network_ready_cb,
+	node_create(sender, app_path, uuid, create_node_ready_cb,
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c]: Why change the order of arguments?

File: mesh/mesh.c [hunk 15]
```diff
@@ -758,12 +759,56 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
...
+		return dbus_error(msg, MESH_ERROR_INVALID_ARGS, "Bad dev UUID");
+
+	if (node_find_by_uuid(uuid))
+		return dbus_error(msg, MESH_ERROR_ALREADY_EXISTS,
+							"Node already exists");
+
+	jnode = json_tokener_parse(json_data);
+
+	sender = l_dbus_message_get_sender(msg);
+	pending_msg = l_dbus_message_ref(msg);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c:682]: If `node_import` fails, there is noone to `unref` the message, and you leak memory.

File: mesh/model.c [hunk 16]
```diff
@@ -228,6 +228,7 @@ static struct l_dbus_message *create_config_update_msg(struct mesh_node *node,
 
 	if ((id & VENDOR_ID_MASK) != VENDOR_ID_MASK) {
 		uint16_t vendor = id >> 16;
+
```
Comment (michallowasrzechonek-silvair) [inline mesh/model.c]: No need for blank lines here, you can leave the whole file unmodified.

File: mesh/node.c [hunk 17]
```diff
@@ -27,6 +27,9 @@
 
 #include <ell/ell.h>
 #include <json-c/json.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Leftover includes

File: mesh/node.c [hunk 18]
```diff
@@ -61,6 +64,7 @@
 #define REQUEST_TYPE_JOIN 0
 #define REQUEST_TYPE_ATTACH 1
 #define REQUEST_TYPE_CREATE 2
+#define REQUEST_TYPE_IMPORT 3
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Maybe make this an enum?

File: mesh/node.c [hunk 19]
```diff
@@ -848,136 +863,6 @@ uint16_t node_generate_comp(struct mesh_node *node, uint8_t *buf, uint16_t sz)
 }
 
 
-#define MIN_COMPOSITION_LEN 16
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: This belongs to a separate patch

File: mesh/node.h [hunk 20]
```diff
@@ -89,8 +89,10 @@ int node_attach(const char *app_path, const char *sender, uint64_t token,
 					node_ready_func_t cb, void *user_data);
 void node_build_attach_reply(struct mesh_node *node,
 						struct l_dbus_message *reply);
-void node_create(const char *app_path, const char *sender, const uint8_t *uuid,
+void node_create(const char *sender, const char *app_path, const uint8_t *uuid,
 					node_ready_func_t cb, void *user_data);
+bool node_import(const char *app_path, const char *sender, void *jnode,
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.h]: Please keep arguments in the same order.

File: doc/mesh-api.txt (hunk 13 above)
Comment (michallowasrzechonek-silvair) [reply doc/mesh-api.txt]: Also, add documentation about `json_data` schema

File: mesh/node.c [hunk 21]
```diff
@@ -1547,11 +1507,49 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
...
+			goto fail;
+
+		if (!keyring_put_net_key(node, PRIMARY_NET_IDX, &net_key))
+			goto fail;
+
+		cb(import_data->user_data, MESH_ERROR_NONE, node);
+
 	} else {
 		/* Callback for create node request */
+		uint8_t dev_key[16];
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Please don't reorder lines if you don't touch the logic. As a rule of thumb, try to keep the patch as small as possible.

File: doc/mesh-api.txt [hunk 22]
```diff
@@ -151,16 +151,34 @@ Methods:
...
+			  provisioning and composition data
+
+		The format must conform to the schema defined in
+		"Mesh Node Configuration Examples" section. Any included token will
+		be ignored in favor of a locally generated token value. If
+		import_data contains composition data (determined by the presence of
+		Elements) it is validated against composition data provided by
+		the application. Otherwise, new node is created based on
+		composition data provided by the application using provisioning data
+		from import_data parameter.
```
Comment (michallowasrzechonek-silvair) [inline doc/mesh-api.txt]: Mention that sequenceNumber is optional

File: doc/mesh-api.txt [hunk 23]
```diff
@@ -151,16 +151,34 @@ Methods:
...
+
+		The data_type parameter defines the import_data type. Supported
+		data_type parameters:
+			- “json”
+
+		The import_data parameter can be either:
+			- Simplified representation of node configuration with
+			  provisioning data only
+			- FULL representation of node configuration with both
+			  provisioning and composition data
```
Comment (michallowasrzechonek-silvair) [inline doc/mesh-api.txt]: "FULL" -> "Full"

File: doc/mesh-api.txt [hunk 24]
```diff
@@ -1061,6 +1079,68 @@ Properties:
...
 		information (e.g., a public key)
 
-Mesh Node Configuration Schema
-==============================
-<TBD>
+Mesh Node Configuration Examples
+================================
+Example of Json format for ImportLocalNode():
+
+	Import local node operation:
```
Comment (michallowasrzechonek-silvair) [inline doc/mesh-api.txt]: Keep it consistent: "Simplified" and "Full" representation, docs for the API don't mention any "migration"

File: mesh/mesh-db.h [hunk 25]
```diff
@@ -16,6 +16,12 @@
  *  Lesser General Public License for more details.
  *
  */
+#include <json-c/json.h>
+
+#define mesh_db_import_node_foreach(obj, key, val) \
+				json_object_object_foreach(obj, key, val)
+
+typedef json_object parsed_type;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.h]: Don't typedef.

File: mesh/mesh-db.h [hunk 26]
```diff
@@ -93,6 +99,10 @@ struct mesh_db_prov {
 	uint8_t priv_key[32];
 };
 
+struct mesh_db_import_node {
+	parsed_type *parsed_node;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.h]: That's a big :-1:. The whole idea smells - you should rather create two separate `mesh_db_import_data` structs, one for simplified flavour and the other for full import, than pass this between `mesh`, `node` and `mesh-db` modules.

Perhaps it would be even possible to reuse `mesh_db_node` for that purpose ("inherit" from it, maybe?).

Comment (michallowasrzechonek-silvair) [reply mesh/mesh-db.h]: This way you wouldn't need to expose that much details in `mesh-db.h`

File: mesh/mesh-db.c [hunk 27]
```diff
@@ -253,6 +253,134 @@ static json_object *jarray_key_del(json_object *jarray, int16_t idx)
...
+	if (json_object_object_get_ex(jobj, "sequenceNumber", &jvalue))
+		return (uint32_t) json_object_get_int(jvalue);
+	else
+		return default_seq_nr;
+}
+
+static bool json_iter(json_object *jobj, struct mesh_db_import_node *i_node)
+{
+	bool dev_k = false, unicast = false;
+	bool ivi = false, ivu = false, net_k = false;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.c]: Please expand names. For example `ivi` is a different entity than `iv_index`, this is confusing!

File: mesh/mesh-db.c [hunk 28]
```diff
@@ -253,6 +253,134 @@ static json_object *jarray_key_del(json_object *jarray, int16_t idx)
...
+				return false;
+			}
+
+		} else {
+			if (strcmp(key, "sequenceNumber"))
+				return false;
+		}
+	}
+
+	if (dev_k && unicast && net_k && ivi && ivu)
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.c]: `return dev_k && unicast && net_k && ivi && ivu`

File: mesh/mesh-db.c [hunk 29]
```diff
@@ -1783,3 +1912,34 @@ bool mesh_db_model_sub_del_all(json_object *jnode, uint16_t addr,
 
 	return delete_model_property(jnode, addr, mod_id, vendor, "subscribe");
 }
+
+
+
+struct mesh_db_import_node *mesh_db_parse_import_data(const char *import_data)
+{
+	json_object *jobj = json_tokener_parse(import_data);
+	struct mesh_db_import_node *i_node =
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.c]: Rename to `import_node`

File: mesh/mesh-db.c [hunk 30]
```diff
@@ -1783,3 +1912,34 @@ bool mesh_db_model_sub_del_all(json_object *jnode, uint16_t addr,
...
+		return NULL;
+	}
+
+	i_node->node->seq_number =
+			parse_imported_sequence(jobj, DEFAULT_SEQUENCE_NUMBER);
+	return i_node;
+
+}
+
+bool mesh_db_import_node_is_simplified(struct mesh_db_import_node *i_node)
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.c]: Remove this function, check for `import_node->db_node->elements` directly

File: mesh/mesh-db.h [hunk 31]
```diff
@@ -93,6 +94,16 @@ struct mesh_db_prov {
 	uint8_t priv_key[32];
 };
 
+struct mesh_db_import_node {
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.h]: Rename to `mesh_db_import`

File: mesh/mesh-db.h [hunk 32]
```diff
@@ -93,6 +94,16 @@ struct mesh_db_prov {
...
+
+	struct mesh_db_node *node;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.h]: Rename to `db_node`

File: mesh/mesh-db.h [hunk 33]
```diff
@@ -93,6 +94,16 @@ struct mesh_db_prov {
...
+	uint8_t dev_key[16];
+	uint8_t net_key[16];
+	uint16_t net_idx;
+	uint32_t iv_idx;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.h]: Rename to 'iv_index'

File: mesh/mesh.c [hunk 34]
```diff
@@ -70,6 +72,10 @@ struct join_data{
 	uint8_t *uuid;
 };
 
+static const char * const supported_import_data_types[] = {
+	"json"
+};
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c]: :+1:

File: mesh/mesh.c [hunk 35]
```diff
@@ -383,6 +389,18 @@ static void node_init_cb(struct mesh_node *node, struct mesh_agent *agent)
 	free_pending_join_call(true);
 }
 
+static bool validate_data_type(const char *data_type)
+{
+	uint8_t len = sizeof(supported_import_data_types) /
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c]: `L_ARRAY_SIZE`

File: mesh/mesh.c [hunk 36]
```diff
@@ -383,6 +389,18 @@ static void node_init_cb(struct mesh_node *node, struct mesh_agent *agent)
...
+					sizeof(supported_import_data_types[0]);
+
+	for (uint8_t idx = 0; idx < len; idx++) {
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c]: Move declaration outside of `for`

File: mesh/mesh.c [hunk 37]
```diff
@@ -593,12 +611,73 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
...
+							"Node already exists");
+
+	if (!l_dbus_message_iter_get_fixed_array(&iter_import_data,
+							&import_data, &n))
+		return dbus_error(msg, MESH_ERROR_INVALID_ARGS,
+							"Bad import_data");
+
+	l_info("PARSE IMPORTED DATA");
+	p_node = mesh_db_parse_import_data(import_data);
+	l_info("PARSE IMPORTED DATA FINISHED");
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c]: Remove logs.

File: mesh/node.c [hunk 38]
```diff
@@ -110,7 +112,13 @@ struct managed_obj_request {
 	void *data;
 	void *cb;
 	void *user_data;
-	uint8_t type;
+	enum request_type type;
+};
+
+struct node_import_request {
+	uint8_t uuid[16];
+	struct mesh_db_import_node *i_node;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: rename `mesh_db_import_node` to `mesh_db_import` and `i_node` to `import`

File: mesh/node.c [hunk 39]
```diff
@@ -870,7 +878,7 @@ uint16_t node_generate_comp(struct mesh_node *node, uint8_t *buf, uint16_t sz)
 #define MIN_COMPOSITION_LEN 16
 
 bool node_parse_composition(struct mesh_node *node, uint8_t *data,
-								uint16_t len)
+			    uint16_t len)
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Ident?

File: mesh/node.c [hunk 40]
```diff
@@ -965,7 +973,7 @@ bool node_parse_composition(struct mesh_node *node, uint8_t *data,
 			vendor_id = l_get_le16(data);
 			mod_id |= (vendor_id << 16);
 			mod = mesh_model_vendor_new(ele->idx, vendor_id,
-									mod_id);
+						    mod_id);
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Indent?

File: mesh/node.c [hunk 41]
```diff
@@ -996,7 +1004,6 @@ bool node_parse_composition(struct mesh_node *node, uint8_t *data,
 
 	return false;
 }
-
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Blank line?

File: mesh/node.c [hunk 42]
```diff
@@ -1566,6 +1565,48 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 
 		cb(node, agent);
 
+	} else if (req->type == REQUEST_TYPE_IMPORT) {
+
+		node_ready_func_t cb = req->cb;
+		struct node_import_request *import_data = req->user_data;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Extract `mesh_db_import` to a local variable, call to `add_local_node` will be shorter.

File: mesh/node.c [hunk 43]
```diff
@@ -1614,14 +1655,22 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
...
 			node_join_ready_func_t cb = req->cb;
 
 			cb(NULL, NULL);
+
+		} else if (req->type == REQUEST_TYPE_IMPORT) {
+			struct node_import_request *import_data =
+								req->user_data;
+			node_ready_func_t cb = req->cb;
+
+			cb(import_data->user_data, MESH_ERROR_FAILED, NULL);
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Other request types call the callback with `req->user_data`, please keep this consistent here.

File: mesh/node.c [hunk 44]
```diff
@@ -1686,6 +1735,50 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
 					req, l_free);
 }
 
+static bool node_import_full(const char *app_path, const char *sender,
+					void *parsed_node, const uint8_t *uuid,
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Rename `parsed_node` to `import` and use real type instead of `void *`.

File: mesh/node.c [hunk 45]
```diff
@@ -1686,6 +1735,50 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
...
+					node_ready_func_t cb, void *user_data)
+{
+	l_debug("");
+	return false;
+}
+
+static bool node_import_simplified(const char *app_path, const char *sender,
+					void *parsed_node, const uint8_t *uuid,
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Rename `parsed_node` to `import` and use real type instead of `void *`.

File: mesh/node.c [hunk 46]
```diff
@@ -1686,6 +1735,50 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
...
+						L_DBUS_INTERFACE_OBJECT_MANAGER,
+						"GetManagedObjects", NULL,
+						get_managed_objects_cb,
+						req, l_free);
+	return true;
+}
+
+bool node_import(const char *app_path, const char *sender, void *parsed_node,
+		const uint8_t *uuid, node_ready_func_t cb, void *user_data)
+{
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Rename `parsed_node` to `import` and cast to real type before passing to `_simplified`/`_full`

**Comment by michallowasrzechonek-silvair** [comment]
Way better :)

File: mesh/node.c (hunk 43 above)
Comment (jakubwitowski-silvair) [reply mesh/node.c]: Won't be done :)

File: mesh/mesh-db.c [hunk 47]
```diff
@@ -1783,3 +1909,31 @@ bool mesh_db_model_sub_del_all(json_object *jnode, uint16_t addr,
...
+
+	if (!jobj) {
+		l_free(import_node);
+		return NULL;
+	}
+
+	import_node->node = l_new(struct mesh_db_node, 1);
+
+	if (!json_iter(jobj, import_node)) {
+		json_object_put(jobj);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-db.c]: IMO lepiej byłoby zrobić `goto error` i tam wyczyścić

Comment (michallowasrzechonek-silvair) [reply mesh/mesh-db.c]: Tak samo przy `if (!jobj)`

File: mesh/mesh.c [hunk 48]
```diff
@@ -593,12 +611,72 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
...
+
+	if (!pending_queue)
+		pending_queue = l_queue_new();
+
+	l_queue_push_tail(pending_queue, pending_msg);
+
+	if (!node_import(app_path, sender, p_node, uuid, create_node_ready_cb,
+								pending_msg)) {
+		l_free(p_node);
+		l_dbus_message_unref(msg);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh.c:698]: Robisz `unref`, a na kolejce `pending_queue` zostaje dangling pointer.

File: mesh/node.c [hunk 49]
```diff
@@ -26,7 +26,7 @@
 #include <sys/time.h>
 
 #include <ell/ell.h>
-#include <json-c/json.h>
+#include <stdio.h>
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Remove

File: mesh/node.c [hunk 50]
```diff
@@ -110,7 +112,13 @@ struct managed_obj_request {
 	void *data;
 	void *cb;
 	void *user_data;
-	uint8_t type;
+	enum request_type type;
+};
+
+struct node_import_request {
+	uint8_t uuid[16];
+	struct mesh_db_import *i_node;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Zmień nazwę pola na `import`

File: mesh/node.c [hunk 51]
```diff
@@ -1566,6 +1565,49 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 
 		cb(node, agent);
 
+	} else if (req->type == REQUEST_TYPE_IMPORT) {
+
+		node_ready_func_t cb = req->cb;
+		struct node_import_request *import_data = req->user_data;
+		struct mesh_db_import *db_node = import_data->i_node;
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Nazwy!

```
struct node_import_request *import_request = req->user_data;
struct mesh_db_import *import = import_request->import;
```

File: mesh/node.c [hunk 52]
```diff
@@ -1686,6 +1736,53 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
 					req, l_free);
 }
 
+static bool node_import_full(const char *app_path, const char *sender,
+					struct mesh_db_import *import_node,
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Samo `import` a nie `import_node` ;P

File: mesh/mesh.c (hunk 48 above)
Comment (michallowasrzechonek-silvair) [reply mesh/mesh.c]: `goto`!
